### PR TITLE
Support for delete edge from diagram tool

### PR DESCRIPTION
**Pull request by hmarchadour**
### Type of this PR 

- [ ] Bug fix
- [x] Improvement
- [ ] Documentation
- [ ] Cleanup
- [ ] Test

### Issue(s)
 
 #119

### What does this PR do?

Call the delete handler on an edge when the remove tool is triggered.

### Screenshot/screencast of this PR

cap 114

### Potential side effects

* Diagram refresh
* Delete tool

### How to test this PR?

- [ ] Frontend Unit tests
- [ ] Backend Unit tests
- [x] Cypress : 
* A test case that delete all edges in Robot topography diagram
- [x] Manual Test : 
* On Robot sample, 
  * create a topography diagram
  * select an edge
  * use the palette delete tool 
    * **we expect a diagram refresh without this edge**
  * select another edge
  * use the [SUPPR] keyboard
    * **we expect a diagram refresh without this edge**

### Checklist

- [x] I have read CONTRIBUTING carefully.
- [x] I have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All my commits are signed-off (`-s`) with my mail address of my Eclipse Account.
- [x] I have covered my changes by unit tests or integration tests or manual tests.
- [x] All tests pass.
- [x] I have updated the documentation accordingly
- [x] New React components are available in storybook

### CODE REVIEW

File: backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/handlers/DeleteFromDiagramEventHandler.java [hunk 1]
```diff
@@ -86,36 +91,104 @@ public boolean canHandle(IDiagramInput diagramInput) {
...
-                this.invokeDeleteTool(node, editingContext, diagram);
-                return new EventHandlerResponse(true, representation -> true, new DeleteFromDiagramSuccessPayload(diagram));
+    private EventHandlerResponse handleDelete(IEditingContext editingContext, Diagram diagram, DeleteFromDiagramInput diagramInput) {
+        List<String> errors = new ArrayList<>();
+        boolean atLeastOneOk = false;
+        for (String edgeId : diagramInput.getEdgeIds()) {
+            var optionalElement = this.diagramService.findEdgeById(diagram, edgeId);
+            if (optionalElement.isPresent()) {
+                Status status = this.invokeDeleteEdgeTool(optionalElement.get(), editingContext, diagram);
+                if (Objects.equals(Status.OK, status)) {
```
Comment (pcdavid) [inline backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/handlers/DeleteFromDiagramEventHandler.java:111]: `Status` is an enum, equality can safely be checked with `==`.

**Review by pcdavid — CHANGES_REQUESTED** [review]

**Comment by pcdavid** [comment]
Except for one, most of the remarks are mostly about style issues, but I have not tested the patch yet. I'all add more remarks if I find issues when testing (not sure when I'll be able to).

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java [hunk 2]
```diff
@@ -22,17 +23,20 @@
...
 @GraphQLInputObjectType
 public final class DeleteFromDiagramInput implements IDiagramInput {
     private UUID projectId;
 
     private UUID representationId;
 
-    private String diagramElementId;
+    private List<String> nodeIds;
+
+    private List<String> edgeIds;
```
Comment (sbegaudeau) [inline backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:39]: Why not List of UUID?

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java [hunk 3]
```diff
@@ -49,16 +53,21 @@ public UUID getRepresentationId() {
         return this.representationId;
     }
 
-    @GraphQLID
     @GraphQLField
     @GraphQLNonNull
-    public String getDiagramElementId() {
-        return this.diagramElementId;
+    public List<@GraphQLNonNull String> getNodeIds() {
```
Comment (sbegaudeau) [inline backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:58]: This should be `List<@GraphQLNonNull @GraphQLID String>` or something like that

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java [hunk 4]
```diff
@@ -49,16 +53,21 @@ public UUID getRepresentationId() {
...
+        return this.nodeIds;
+    }
+
+    @GraphQLField
+    @GraphQLNonNull
+    public List<@GraphQLNonNull String> getEdgeIds() {
```
Comment (sbegaudeau) [inline backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:64]: This should be `List<@GraphQLNonNull @GraphQLID String>` or something like that

File: backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/handlers/DeleteFromDiagramEventHandler.java [hunk 5]
```diff
@@ -86,36 +91,104 @@ public boolean canHandle(IDiagramInput diagramInput) {
     public EventHandlerResponse handle(IEditingContext editingContext, Diagram diagram, IDiagramInput diagramInput) {
         this.counter.increment();
 
+        final EventHandlerResponse result;
```
Comment (sbegaudeau) [inline backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/handlers/DeleteFromDiagramEventHandler.java:94]: No need to mark this final

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: frontend/src/diagram/DiagramWebSocketContainer.tsx [hunk 6]
```diff
@@ -299,16 +299,40 @@ export const DiagramWebSocketContainer = ({ representationId, selection, setSele
...
+       */
+      diagramElements
+        .filter((diagramElement) => diagramElement instanceof SNode)
+        .filter((diagramElement) => diagramElement.parent instanceof SNode)
+        .forEach((diagramElement) => {
+          const index = nodeIds.indexOf(diagramElement.id);
+          if (index > -1 && nodeIds[index] !== diagramElement.id) {
+            nodeIds.splice(index, 1);
+          }
+        });
```
[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/src/diagram/DiagramWebSocketContainer.tsx:325]: You are totally right.
This piece of code is not testable without a multi selection use case.
I think I should move it in another PR and let 
`nodeIds = diagramElements.filter(diagramElement => diagramElement instanceof SNode).map(elt => elt.id)`
here for now.

A manual(or cypress) test case with this scenario could be great, doesn't it?

[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/src/diagram/DiagramWebSocketContainer.tsx:325]: Or simply make the `deleteElement` to `deleteElements` in the multi selection PR (but it become tedious to move...)

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java (hunk 3 above)
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:58]: I must update the GraphQLID annotation in this purpose but you are right, I should make this precision.

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java (hunk 2 above)
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:39]: This implies a GraphQLID update that I've not think todo.

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java (hunk 4 above)
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:64]: This comment is interesting.
You say that an edgeId is an UUID but the getIdProvider return a String.
If I see the EdgeMappingConverter, the idProvider give a string ``` sourceId + " --> " + targetId + " - " + count```
(It's not a disguised UUID)

This @GraphQLID implies :
* Edge#id
* EdgeDescription#idProvider
* EdgeElementProps
* IDiagramService#findEdgeById

This update should be done in a dedicated enhancement.

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java (hunk 3 above)
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:58]: This comment is interesting.
You say that an nodeId is an UUID but the getIdProvider return a String.
If I see the NodeMappingConverter, the idProvider give a string ```UUID.randomUUID().toString()```
(It's a disguised UUID) 

Actually, the UUID implementation of a node ID does not appear in Types :
* Node#id
* NodeDescription#idProvider
* NodeElementProps 
* IDiagramService#findNodeById 

I think that this ```@GraphQLID``` should be add in a dedicated enhancement.

File: frontend/src/diagram/DiagramWebSocketContainer.tsx [hunk 7]
```diff
@@ -299,16 +299,40 @@ export const DiagramWebSocketContainer = ({ representationId, selection, setSele
...
+      diagramElements.forEach((diagramElement) => {
+        if (diagramElement instanceof SNode) {
+          nodeIds.push(diagramElement.id);
+        } else if (diagramElement instanceof SEdge) {
+          edgeIds.push(diagramElement.id);
+        }
+      });
+      /**
+       * We should not ask to remove more than once a Node.
+       * To avoid this state, we filter all nodes that have also there parents in the list to delete.
```
[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/src/diagram/DiagramWebSocketContainer.tsx]: I've remove this part on the following commit (I keep this feedback for the multi-selection)

File: frontend/src/diagram/DiagramWebSocketContainer.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/src/diagram/DiagramWebSocketContainer.tsx]: I've remove this part on the following commit (I keep this feedback for the multi-selection)

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java (hunk 2 above)
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/dto/DeleteFromDiagramInput.java:39]: @see following conversation

File: frontend/src/diagram/DiagramWebSocketContainer.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (pcdavid) [reply frontend/src/diagram/DiagramWebSocketContainer.tsx]: The function itself could be extracted and generalized a little to be testable: "given a set of element and a function which computes the parent/container of an element in the set, return the minimum subset which covers all the input elements".

```js
function minimize(elements, parentFn) {
   // ...
}
```

**Comment by pcdavid** [comment]
@sbegaudeau It seem that Hugo has taken your remarks into account (I have not looked at the changes, but he has pushed new commits since your review). I've tested the patch, it seems to work fine. Should we merge it?

**Review by sbegaudeau — APPROVED** [review]
Everything is fine for me, could you rebase it on top of master to trigger the build with the latest changes to see if there are any issue?

**Comment by hmarchadour** [comment]
I've rebased this PR. Manual/auto tests seems ok.

**Review by pcdavid — APPROVED** [review]

**Pull request merged**